packages/rolldown/src/utils/compose-js-plugins.ts

```diff
@@ -0,0 +1,276 @@
+          batchedHooks.load = handlers
+          if (plugin.load) {
+            handlers.push(plugin.load)
+          }
+          break
+        }
+        case 'transform': {
+          const handlers = batchedHooks.transform ?? []
+          batchedHooks.transform = handlers
+          if (plugin.transform) {
+            handlers.push(plugin.transform)
+          }
+          break
+        }
+        case 'resolveId': {
+          const handlers = batchedHooks.resolveId ?? []
+          batchedHooks.resolveId = handlers
+          if (plugin.resolveId) {
+            handlers.push(plugin.resolveId)
+          }
+          break
+        }
+        case 'buildEnd': {
+          const handlers = batchedHooks.buildEnd ?? []
+          batchedHooks.buildEnd = handlers
+          if (plugin.buildEnd) {
+            handlers.push(plugin.buildEnd)
+          }
+          break
+        }
+        case 'renderChunk': {
+          const handlers = batchedHooks.renderChunk ?? []
+          batchedHooks.renderChunk = handlers
+          if (plugin.renderChunk) {
+            handlers.push(plugin.renderChunk)
+          }
+          break
+        }
+        case 'augmentChunkHash':
+        case 'banner':
+        case 'generateBundle':
+        case 'moduleParsed':
+        case 'onLog':
+        case 'options':
+        case 'outputOptions':
+        case 'renderError':
+        case 'renderStart':
+        case 'resolveDynamicImport':
+        case 'writeBundle': {
+          throw new Error(
+            `Failed to compose js plugins. Plugin ${pluginName} has an unsupported hook: ${pluginProp}`,
+          )
+          break
+        }
+        default: {
+          // All known hooks should be handled above. We allow plugin to have unknown properties and we just ignore them.
+          const _executiveCheck: never = pluginProp
+        }
+      }
+    })
+  })
+
+  const composed: Plugin = {
+    name: `Composed(${names.join(', ')})`,
+  }
+
+  ;(Object.keys(batchedHooks) as (keyof typeof batchedHooks)[]).forEach(
+    (hookName) => {
+      switch (hookName) {
+        case 'buildStart': {
+          if (batchedHooks.buildStart) {
+            const batchedHandlers = batchedHooks.buildStart
+            composed.buildStart = async function (options) {
+              await Promise.all(
+                batchedHandlers.map((handler) => {
+                  const [handlerFn, _handlerOptions] = normalizeHook(handler)
+                  return handlerFn.call(this, options)
+                }),
+              )
+            }
+          }
+          break
+        }
+        case 'load': {
+          if (batchedHooks.load) {
+            const batchedHandlers = batchedHooks.load
+            composed.load = async function (id) {
+              for (const handler of batchedHandlers) {
+                const [handlerFn, _handlerOptions] = normalizeHook(handler)
+                const result = await handlerFn.call(this, id)
```

Comment:
We should check the return value of `handlerFn` and only await when it is actually a Promise. The extra `await` tick is actually costly in hot code paths. This applies to all serial hooks.

Vue SSR got a nice perf boost by only `await` when necessary: https://github.com/vuejs/core/commit/05779a70bd0b567ae458a07636d229bd07c44c4e

Reply:
That's an very nice idea. I'll look into it.